### PR TITLE
Raise error if KEYSPACE has invalid characters

### PR DESCRIPTION
**Pull request by dexter816**

### CODE REVIEW

File: docs/deployment.md [hunk 1]
```diff
@@ -13,52 +13,44 @@ There are orchestration templates for running Jaeger with:
...
 
-In the future we will support different service discovery systems to dynamically load balance
-across several collectors ([issue 213](https://github.com/uber/jaeger/issues/213)).
+For production deployment, pass `MODE=prod DATACENTER={datacenter}` arguments to the script,
+where `{datacenter}` is the name used in the Cassandra configuration / network topology.
+
+The script also allows overriding TTL, keyspace name, replication factor, etc.
+Run the script without arguments to see the full list of recognized parameters.
+
+### ElasticSearch
```
Comment (jpkrohling) [inline docs/deployment.md]: The "s" in Elasticsearch should be in lower case. Yeah, I think it's weird too :)

File: docs/deployment.md [hunk 2]
```diff
@@ -77,6 +70,12 @@ or, if you don't have the source code
 docker run -it --rm jaegertracing/jaeger-collector /go/bin/collector-linux -h
 ```
 
+Example:
+
+```
+docker run -it --rm -p14267:14267 -p14268:14268 jaegertracing/jaeger-collector /go/bin/collector-linux --cassandra.keyspace=jaeger_v1_test --cassandra.servers=192.168.0.183
```
Comment (jpkrohling) [inline docs/deployment.md]: It might be more readable if you break the lines, like this:

```
docker run -it \
  --rm \
  --name jaeger-collector \
  -p14267:14267 \
  -p14268:14268 \
  --link cassandra:cassandra \
  jaegertracing/jaeger-collector \
  /go/bin/collector-linux --cassandra.keyspace=jaeger_v1_local --cassandra.servers=cassandra
```

Or, using env vars:

```
docker run -it \
  --rm \
  --name jaeger-collector \
  -p14267:14267 \
  -p14268:14268 \
  -e CASSANDRA_KEYSPACE=jaeger_v1_local \
  -e CASSANDRA_SERVERS=cassandra \
  --link cassandra:cassandra \
  jaegertracing/jaeger-collector
```

(note that I changed the keyspace to be consistent with the previous command that starts Cassandra, and added a link to that container, so that we don't need IPs that won't work for the reader)

File: docs/deployment.md [hunk 3]
```diff
@@ -85,44 +84,65 @@ Port  | Protocol | Function
...
 
 ## Query Service & UI
 
 **jaeger-query** serves the API endpoints and a React/Javascript UI.
 The service is stateless and is typically run behind a load balancer, e.g. nginx.
 
+An example to test Query Service:
+
+```
+docker run -it -p16686:16686 jaegertracing/jaeger-query:latest /go/bin/query-linux --cassandra.keyspace=jaeger_v1_test --cassandra.servers=192.168.0.183
```
Comment (jpkrohling) [inline docs/deployment.md]: Break this one too, like my previous comment.

Comment (dexter816) [reply docs/deployment.md]: Sure :+1: on it :)

**Comment by dexter816** [comment]
@jpkrohling thanks for the review , I made two seperate commits for issue #472 and #445 , and squashed other commits could this be merged :)

File: plugin/storage/cassandra/schema/create.sh [hunk 4]
```diff
@@ -41,7 +41,8 @@ else
 fi
 
 keyspace=${KEYSPACE:-"jaeger_v1_${datacenter}"}
-
+datacenter= $(echo $datacenter| sed 's/[-.]/_/g')
```
Comment (yurishkuro) [inline plugin/storage/cassandra/schema/create.sh]: don't need to change data center, it's not used as free-form identifier

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -125,7 +125,7 @@ If you are adding a new file it should have a header like below.  The easiest
 way to add such header is to run `make fmt`.
 
 ```
-// Copyright (c) 2017 The Jaeger Authors.
+// Copyright (c) 2017 Uber Technologies, Inc.
```
Comment (yurishkuro) [inline CONTRIBUTING.md]: these changes don't belong to this PR, please sync the master and rebase your branch off of it

File: plugin/storage/cassandra/schema/create.sh [hunk 6]
```diff
@@ -42,6 +42,9 @@ fi
 
 keyspace=${KEYSPACE:-"jaeger_v1_${datacenter}"}
 
+#replaces hyphen and period with underscore
+keyspace=$(echo $keyspace | sed 's/[-.]/_/g')
```
Comment (yurishkuro) [inline plugin/storage/cassandra/schema/create.sh]: I don't think this is the best fix for the original problem. For example, I can still give a weird datacenter name like `dc#1` and the CQL will blow up on invalid keyspace name.  A better solution is to simply check for invalid characters (via exclusion, i.e. `[^a-zA-z0-9_]`) and throw an error, advising the user to make their own decision and pass the keyspace as KEYSPACE env var.

**Comment by yurishkuro** [comment]
```
$ MODE=prod DATACENTER=a1_2 ./plugin/storage/cassandra/schema/create.sh
Error: invalid characters in KEYSPACE parameter, please use letters, digits or underscores
```

File: plugin/storage/cassandra/schema/create.sh [hunk 7]
```diff
@@ -42,6 +42,9 @@ fi
 
 keyspace=${KEYSPACE:-"jaeger_v1_${datacenter}"}
 
+valid_char='a-zA-Z0-9_'
+if [[ $keyspace =~ ^[$valid_char]* ]]; then usage "invalid characters in KEYSPACE parameter, please use letters, digits or underscores"; fi 
+
```
Comment (yurishkuro) [inline plugin/storage/cassandra/schema/create.sh]: this works
```
if [[ $keyspace =~ [^a-zA-Z0-9_] ]]; then
   usage "invalid characters in KEYSPACE parameter, please use letters, digits or underscores"
fi
```

Comment (yurishkuro) [reply plugin/storage/cassandra/schema/create.sh]: It's better to print the value too
```
usage "invalid characters in KEYSPACE=$keyspace parameter, please use letters, digits or underscores"
```

**Comment by yurishkuro** [comment]
Resolves #445

**Comment by dexter816** [comment]
Hey I didn't intend to close this pull request but seems like I cannot reopen the PR , I am working on understanding git workflow ,but can you help reopen this request , thanks

**Comment by yurishkuro** [comment]
I can't open it either, maybe because it's the second time you closed it. Just open another one, it doesn't matter.
